core/evm.go

```diff
@@ -501,3 +515,137 @@ func Transfer(db vm.StateDB, sender, recipient common.Address, amount *big.Int,
 		db.AddBalance(recipient, amount)
 	}
 }
+
+// FetchStakingInfoFn is the EVM implementation of the read only staking precompile
+// avoid import cycle by putting this function here
+func FetchStakingInfoFn(ref *block.Header, chain ChainContext) vm.RoStakingPrecompileFunc {
+	return func(db vm.StateDB, roStakeMsg *stakingTypes.ReadOnlyStakeMsg) ([]byte, error) {
+		switch roStakeMsg.What {
+		case "DelegationByDelegatorAndValidator":
+			{
+				// either we iterate over ReadDelegationsByDelegator
+				// or over wrapper.Delegations. since ReadDelegationsByDelegator
+				// is only written at the end of a block, prefer wrapper.Delegations
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				for _, delegation := range wrapper.Delegations {
```

Comment:
This seems too costly. Why not use the delegation index?

Reply:
This one was from me. 
- Even with the `DelegationIndex`, we will have to iterate over each index to find the relevant validator with which the address matches.
- Delegation indices are written at the end of each block and contracts which check this during a transaction may find the number to be inaccurate, if the delegation is new.